raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -317,6 +317,9 @@ private void appendLeaderChangeMessage(LeaderState state, long currentTimeMs) {
             .map(follower -> new Voter().setVoterId(follower))
             .collect(Collectors.toList());
 
+        // Adding the leader to the voters as the protocol ensures that leader always votes for itself.
+        voters.add(new Voter().setVoterId(state.election().leaderId()));
```

Comment:
Similar to `LeaderState::nonEndorsingFollower`, I think you want to add a method to `LeaderState` with the following signature `public Set<Integer> endorsingFollower()`.

Reply:
Thanks @jsancio . I am slightly confused regarding your comment though(maybe because of my limited understanding of the protocol) so plz help me understand here:
The block of code where i made. the changes, picks all the followers and adds them as the voters:

`List<Voter> voters = state.followers().stream()
            .map(follower -> new Voter().setVoterId(follower))
            .collect(Collectors.toList());`

While the LeaderChange message schema, states that voterIds are all those voters who voted for the currently elected leader. 

Continuing on this, from what I understood in the handling of BeginQuorumEpoch response handling section in kIP-595, the moment a leader receives a response from a voter with no errors, it records an endorsement from the voter for itself. It also states that if a Voter accepts a BeginQuorumEpoch(and thereby sending the accepted response), it automatically transitions to a follower state for that leader. Lastly, the new leader will keep sending BeginQuorumEpoch requests to non endorsing voters until it doesn't receive them (or doesn't infer through the Fetch API in case of partitioning).

So, the questions I have are:

1. Since the leader waits for the endorsements from all voters, doesn't it mean that eventually the followers method in LeaderState would return all voters? While the LeaderState message schema mentions that VotedIDs are those voters are voted for this new leader. 
2. I had thought that the nonEndorsingFollower method was used to find all followers who haven't yet endorsed for the new leader and keep sending requests to them. In that assumption, I am not able to follow your statement of adding a new method to add endorsingFollower wrt this PR. 

Again, I apologise if I made some comments which don't make sense. As I said, I am not totally well verse with the internal implementations being done here.

Reply:
I think the original KIP stated that the LeaderChange message would encode the set of voters that had voted for the leader. We thought this might be useful for debugging. Later on, we had a change of heart and decided it would just be the set of voters. Now I'm thinking it might be useful to have both. The log will always remember who the voters were at the time of the election and which voters had granted the leader's candidacy, which could be helpful in case of misconfigurations.

For the set of voters which voted for the current leader, I think what we want is `CandidateState.grantingVoters`. However, by the time `onBecomeLeader` is fired, we have already dropped the `CandidateState`. One option is to carry `grantingVoters` over to `LeaderState`. We might also be able to pass it through `onBecomeLeader`. This will be easier if we get rid of the call to `onBecomeLeader` in `initialize()`. Following KAFKA-10527, it is not possible to initialize as a leader, so we could raise an exception instead.

Reply:
Thanks @hachikuji . So, from the context of this PR, do you suggest to continue using the definition of Voters in the LeaderChange message as all the voters ? Maybe we can create separate issues for:
1) changing LeaderChange message to include both all voters and endorsing voters for that Leader.
2) Make relevant changes to be able to pass the grantingVoters information from CandidateState to LeaderState. These include the things you mentioned like removing `onBecomeLeader` from `initialize()` or throw exception.

Just curious on the last part. As per KAFKA-10527, a node can't be initialized as a leader. So, this block of code is effectively unused then:

`if (quorum.isLeader()) {
onBecomeLeader(currentTimeMs);
        } else if (quorum.isCandidate()) {
            onBecomeCandidate(currentTimeMs);
        } `


Let me know how should I proceed here...

Reply:
@vamossagar12 Up to you I guess. I'm ok doing it all here since the changes seem pretty small.

Reply:
Alright. Thanks @hachikuji , I will make the changes and update the PR.

Reply:
@hachikuji i have made the changes you had suggested. 
cc @jsancio

Reply:
hey @hachikuji, did you get a chance to review the changes?